start writing parser

koddsson opened the pull request.

koddsson committed:
start writing parser

keithamus commented on src/expr.rs:
```diff
@@ -0,0 +1,15 @@
+
+pub struct Expr {
+    // I don't fully understand why this works with a `Box<>` and fails without.
+    pub left: Option<Box<Expr>>,
```
Every struct has a size which _must_ be known at runtime, because that's how Rust will know how to allocate memory for that struct. If it can't determine the size, it can't allocate memory for it. So for example if `Token` was an `i8` it would be 1byte, `struct Expr { operator: Token }` would also be 1byte, while `struct Expr {left: Token, right: Token}` would be 2bytes and so on. This is documented [in the rust reference manual, on Type Layout](https://doc.rust-lang.org/reference/type-layout.html).

Now, if a struct references itself, suddenly the size cannot be known, because it has a recursive size; it holds a copy of itself which holds a copy of itself and so on. This means we can't compile that code because there'd be no reasonable way to allocate this struct... in theory it could need infinite memory to store. Even if you could make some rule that says "this will only ever be nested to a depth of X" it would still always need to allocate X every time, which would be an incredibly wasteful use of memory.

The way around self-referencing structs is to store the child struct elsewhere (like the heap), and instead of holding the struct within itself, it holds a _pointer_ to the other memory location. We know the size of a pointer (`usize` - 4bytes on 32bit systems or 8bytes on 64bit systems), and so if we held a pointer instead of the object itself, suddenly we _can_ work out the size at runtime. Rust can now say to its memory allocator "I'd like `usize` bytes to hold this data", and it will fill that data with a different memory location so which will also be `usize` bytes to a different memory location and so on.

The way to do this pointer indireciton in Rust is to use `Box`. To quote [the docs](https://doc.rust-lang.org/std/boxed/struct.Box.html):

> A pointer type that uniquely owns a heap allocation of type T.

So while an `Expr` would be of _some size_, a `Box<Expr>` is always `usize` bytes. The size of `struct Expr { child: Expr }` could never be known (as it's recursive), but `struct Expr { child: Box<Expr> }` has a known size (`usize` bytes (either 4 or 8 bytes depending on what system it's being compiled on)).

keithamus commented on src/expr.rs:
```diff
@@ -0,0 +1,15 @@
+pub struct Expr {
+    // I don't fully understand why this works with a `Box<>` and fails without.
+    pub left: Option<Box<Expr>>,
+    pub operator: Option<&'static Token>,
```
It's odd that this has a `&'static` lifetime. That effectively says that `Token` lives for the entire life of the program, and I'm not sure why that's needed. 

What happens if you remove it entirely?

```suggestion
    pub operator: Option<Token>,
```

koddsson replied:
Yeah this is me just trying stuff from online to solve errors in the compiler, I forget why I needed a lifetime here and I think I started with `'a` but tried `'static` when that didn't work.

I think it's probably just easy to remove.

keithamus commented on src/main.rs:
```diff
@@ -3,8 +3,12 @@ use std::fs;
 use std::process::ExitCode;
 
+use crate::parser::Parser;
 use crate::scanner::Scanner;
```
Btw `Scanner` has a field `source: String`. That's not ideal. In Rust the `String` type is a heap allocated buffer. Every time you copy `String` it will be re-allocated, so you'll store the whole thing again and again and again.

Rust has the `str` type, which acts as a pointer to the originally heap allocated string. Here's where lifetimes get a little tricky because you'll probably end up having to tell Rust that:

```rust
struct Scanner<'the_lifetime_of_the_string> {
  source: &'the_lifetime_of_the_string str, 
}
```

This way you're letting Rust know that `Scanner` can only be around in memory while the String is also around in memory.

koddsson replied:
Right, I figured this was the case but I just wanted this to compile and couldn't be bothered with the lifetimes at the time. I still feel like I don't have the mental modal for lifetimes so I was worried I would create some error that I wouldn't be able to solve.

keithamus replied:
Think about lexical scoping. In JS:

```js
function foo() {
  let bar = {a: 1};
  {
    let baz = {};
    // Do stuff...
  }
  console.log(bar.a)
  // Can't use baz here as it's undefined
  typeof bar == 'undefined'
}

// Can't use bar here, as it's scoped to foo()
typeof bar == 'undefined'
```

Let's look at a psuedo language which had no way of declaring variables and no way handling lexcial scope, instead you had to call a magic `__drop` every time you wanted to free up memory:

```js
function foo() {
  bar = {a: 1};
  baz = {};
  // Do stuff...
  __drop(baz);
  console.log(bar.a)
  __drop(bar);
}
```

This is effectively what lexical scopes do for you. They drop a variable if it leaves a scope. A lexical scope is a _lifetime_. We know `bar` only lives inside of `foo()`. When we finish executing `foo()`, we can delete `bar`. 

However there's a problem. What happens if we give the value to another function:

```js
function foo() {
  bar = {a: 1);
  someOtherFunction(bar);
}
```

We can no longer delete `bar` as we exit `foo`, because we don't know what `someOtherFunction` is going to do to it. It might send it myriad other places, do async stuff or log it to the console or whatever. In JS we have the garbage collector that keeps track of all these, and knows when to delete it. Rust doesn't have a garbage collector, and so it's difficult to know when it should be deleted. In languages like C/C++ you're often left to your own devices on when this stuff gets freed.

Let's use our pseudo language to introduce a bug to illustrate a failure mode in a language without GC:

```js
function foo() {
  bar = {a: 1};  // Step 1
  someOtherFunction(bar); // Step 2
  __drop(bar); // Step 4
}
function someOtherFunction(bar) {
  await oneTick(); // Step 3
  bar.a // Step 5 OOPS! `bar` was dropped and so this is now a segfault
}
```

One way this gets solved in languages without GC is to use `Ref` or `RefPtr` classes which count every time they're moved around. Let's introduce something like that to our pseudo language:

```js
class Ref {
  #owners = 1;
  constructor(value) {
    this.value = value;
  }

  release() {
    this.#owners -= 1;
    if (this.#owners < 1) {
      // Here's our magic drop function again
      __drop(this.value);
    }
  }

  give() {
    this.#owners +=1;
  }
}

function foo() {
  bar = new Ref({a:1}); // Step 1; `Ref.#owners == 1`
  someOtherFunction(bar.give()); // Step 2; `Ref.#owners == 2`
  bar.release(); // Step 4; `Ref.#owners ==1`
}

function someOtherFunction(bar) {
  await oneTick(); // Step 3; `Ref.#owners == 2`
  bar.value.a; // Step 5, bar is still in memory as it has 1 owner!
  bar.release(); // Step 6, `Ref.#owners == 0`, `bar` is dropped.
}
```

Here we can see by keeping a bit of memory around (the counter) and by adding some more code, we can be more safe about how many times our value moves around our code, and only free it when nothing else is referencing it.

However, what if we could alter our language/compiler to do all of this for us? What if we could "annotate" variables to alter their lifetimes, letting the compiler reason about these things for us... Let's change our pseudo language and enforce that every variable, parameter, and function has to declare a "lifetime" and our `__drop` function can drop a whole set of variables that belong to a lifetime:

```ts
// `foo` has an `A` scope/lifetime
function<'A> foo() {
 <'A>bar = {a:1};  // `bar` has to live as long as `A`.
  someOtherFunction(bar);
  __drop(<'A>); // Drop all `'A` lifetimes
}

// `someOtherFunction` has the `B` scope/lifetime, and the `bar` variable _has_ to live as long as `B`.
function<'B> someOtherFunction(<'B>bar) {
  // Opaque function
  __drop(<'B>); // Drop all `'B` lifetimes
}
```

Now the compiler has some new information to think about. It doesn't know what's going on in `someOtherFunction`, but we've told it that `someOtherFunction` has this `B` lifetime which it keeps for its own execution, and we've attached the `B` to `bar`. So the compiler (again, without thinking about the code inside) knows that `someOtherFunction` _has to keep `bar` around long enough for `someOtherFunction` to finish_. And in this way, we now know we can't call `__drop` on `bar` at the end of `foo`, because the lifetime `A` has to live as long as `B` (and vice versa).

So here, the compiler now has some choices:

 - By giving `bar` to `someOtherFunction`, we give it a new lifetime of `'B` and `__drop<'A>()` won't drop `bar`. Instead, `bar` now lives as long as `'B` and gets dropped when `__drop<'B>();` is called.
 - Throw a compile error telling the user that they need to do other special things to `bar` so it can be safely passed to `someOtherFunction`.

Rust takes this a step further, because of borrowing and mutability. However those topics - while adjacent - are different enough that they don't _need_ to be discussed as part of lifetimes.

koddsson replied:
> A lexical scope is a _lifetime_.

> Rust takes this a step further, because of borrowing and mutability. However those topics - while adjacent - are different enough that they don't _need_ to be discussed as part of lifetimes.

These two paragraphs a long with the very excellent code examples make me feel like I understand lifetimes now.

I can probably take that knowledge and refactor the code now to use lifetimes where applicable.

Thanks ✨

koddsson replied:
I've seen `'a` and `'static` annotations been used. Are there conventions and/or keywords for lifetimes? Is `'a` a convention and `'static` and lifetime keyword?

keithamus replied:
Rust has two "special" reserved lifetimes: `'static` which means "the life of the executable". This tells the compiler this variable will always be present in memory. The other is the "placeholder" lifetime `'_` which tells the compiler to "just figure it out".

All other lifetimes are made up. `'a` is kind of like `<T>`. It has become popular lifetime name but it holds no meaning.

koddsson replied:
The scanner is now using `&'a str` for the source code. I might have added more errors but hopefully I'm going in the right direction.

keithamus replied:
Yep `&'a str` sounds perfect.

keithamus commented on src/parser.rs:
```diff
@@ -0,0 +1,196 @@
+                return;
+            }
+
+            let ignored_tokens = vec![
```
Storing this in a vec is a little strange, because you're likely allocating memory here that you needn't. I think you'd be better off using pattern matching (so `if matches!(thing, TokenType::CLASS | TokenType::FUN...`), and maybe putting that into a method on TokenType or something.

keithamus reviewed (commented):
I think there's a ton to cover here but I don't want to overwhelm so I've added a few bits and I can re-review later or we can pair on it or something.

koddsson committed:
Remove a probably unneeded `&'static` lifetime

Co-authored-by: Keith Cirkel <keithamus@users.noreply.github.com>

koddsson committed:
use `matches!` instead of a vector

koddsson committed:
remove comment

koddsson committed:
add a todo function for parse

koddsson commented on src/parser.rs:
```diff
@@ -0,0 +1,200 @@
+        }
+    }
+
+    fn unary(self) -> Unary {
+        if self.match_types(vec![TokenType::BANG, TokenType::MINUS]) {
+            let operator = self.previous();
+            let right = self.unary();
+            return Unary {
+                operator,
+                right: Some(Box::new(right)),
+            };
+        }
```
@keithamus; Any idea on how I should represent the `Expr` -> `Unary` super class -> sub class relationship? The book just subclasses because it's in Java but I'm not sure what the equivalent would be in Rust.

keithamus replied:
The easiest way, if you have a known set of "subclasses" is to group them all into an enum:

```rust
enum Expr {
  Unary { operand: Box<Expr>, operator: Token },
  Binary { left_operand: Box<Expr>, operator: Token, right_operand: Box<Expr> },
}
```

Of course this will require you to match against each type for any method where you might return data, but that's likely okay.

koddsson committed:
add a bunch of lifetimes

koddsson commented on src/parser.rs:
```diff
@@ -0,0 +1,202 @@
+        return expr;
+    }
+
+    fn match_types(&mut self, types: Vec<TokenType>) -> bool {
```
I guess having a vector here as well isn't great? I'm just doing it so I can pass in `n` many items. Could I do some sort of spread operator here instead?

keithamus replied:
I think this perhaps over-abstracts when you could use `if !matches(foo, TokenType::Bar | ...)`.

If you'd really like to avoid the typing, you can either write methods to return the specific match types, e.g...

```rust
fn matches_literal(&self) -> bool {
  matches!(self,  Self::IDENTIFIER | Self::STRING | Self::NUMBER)
}

// Elsewhere:
if token.matches_literal() { ... }
```

Or write a macro... but again that might be over-engineering. At any rate here's how the above function could be a macro instead:

```rust
macro_rules! matches_literal {
  ($val: ident) => {
    matches!($val,  TokenType::IDENTIFIER | TokenType::STRING | TokenType::NUMBER)
  }
}

// Elsewhere:
if matches_literal!(token) { ... }
```

keithamus replied:
> Could I do some sort of spread operator here instead?

Rust explicitly does not have variadic functions. If you want something to be variadic, you can only use macros.

koddsson replied:
> I think this perhaps over-abstracts when you could use `if !matches(foo, TokenType::Bar | ...)`.

How should the `matches` function signature look if I did this?

keithamus replied:
I would look at the call sites that are calling `.match_type(some things)` and turn those into a function like `.is_type_of_things()`.

As an example you have `self.match_types(vec![TokenType::BangEqual, TokenType::EqualEqual])`. You could instead have a function like: 

```rust
#[inline]
fn is_equality_check(&self) -> bool {
  matches!(self, TokenType::BangEqual, TokenType::EqualEqual)
}
```

This way the callsite could use `self.is_equality_check()` instead. Hopefully this would make the codebase more maintainable because if you had this check everywhere, and needed to add something like `TokenType::SortOfEqual` or `TokenType::StrictlyEqual`, there's just one place to add it.

---

As an aside, the compiler can often inline simple functions like these so there's often no performance trade-off of making things functions. Rust will inline functions as part of a PGO (Profile Guided Optimization) pass (typically only enabled for release builds), but you can tell it to do so during LTO (Link Time Optimization) by adding the `#[inline]` annotation atop the function. It won't _always_ inline the function, especially if it's exported. You can instead add `#[inline(always)]` to effectively force Rust to inline it, but generally speaking that's not a super great idea. The downside to inlining is that it can make interactive debugging harder as there won't be a symbol for the function (because it effectively disappears).

koddsson committed:
remove unary for now

koddsson committed:
add a couple of muts?

koddsson committed:
use lifetimes instead of String in scanner

koddsson committed:
fix lifetimes in Parser

koddsson committed:
more work

koddsson committed:
more expression fixes?

koddsson committed:
no errors??

koddsson commented:
I've been looking at https://github.com/tdp2110/crafting-interpreters-rs/ for help and I think I've managed to fix all the borrow errors I wasn't getting.

Not sure I'm still getting why certain errors might occur but I feel like something is started to click!

koddsson commented on src/parser.rs:
```diff
@@ -0,0 +1,242 @@
+    fn equality(&mut self) -> Result<Expr, ParseError> {
+        let mut expr = self.comparison()?;
+        while self.match_types(vec![TokenType::BANG_EQUAL, TokenType::EQUAL_EQUAL]) {
+            let operator = self.previous();
+            let maybe_binary_operator = Parser::token_to_binary_operator(operator);
+
+            let right = self.comparison()?;
```
I was getting really weird borrow errors here. I was looking at my IDE and it wasn't really explaining the issue well. Looking at the terminal output pointed me in the right direction. 

I previously had it like this:

```suggestion
            let operator = self.previous();
            let right = self.comparison()?;
            let maybe_binary_operator = Parser::token_to_binary_operator(operator);

```

And the order really mattered? I should make sure to use the `operator` variable before calling the other method since the other method couldn't take ownership of `self`(?) while `operator` was still being used here in `equality`?

keithamus replied:
If you're borrowing some or all of a struct (e.g. by calling `let right = self.comparison()`) then you can't "give" the struct to a mutable function (I see `previous` is `fn previous(&mut self) -> &Token`). This is because the function which has the mutable reference could change the data from under you. So you've got some options: 

 - Change the call order (like you did) so that mutations happen before reads.
 - Change functions to avoid references, and instead copy or clone data (less memory efficient but frequently worth doing)
 - Try to find ways to avoid `&mut`. It looks like `previous()` doesn't actually _need_ `&mut self` so maybe try changing it to `&self` and see if you can change the order without compile failures.

koddsson committed:
Allow explicit returns

koddsson committed:
use eprintln instead of write to stderr

koddsson committed:
remove unneeded lifetime

koddsson committed:
while is actually if

koddsson committed:
actually implement the parse method

koddsson committed:
fix caseing of token types

koddsson committed:
remove nightly tag from rust fmt

koddsson committed:
fix clippy comments

koddsson committed:
implement Display for a lot of expressions

koddsson committed:
implement more formatters

koddsson committed:
trigger ci

koddsson committed:
trigger ci

koddsson committed:
fuck around to try to fix formatting

koddsson committed:
Revert "fuck around to try to fix formatting"

This reverts commit 815920d42dcc9f225382bb2762a4972626da8fab.

koddsson committed:
fix recursive call to Display impl

koddsson committed:
fix more token display impl

koddsson committed:
maybe fix token_type display?

koddsson committed:
format numbers correctly?

koddsson committed:
output fractionals with at least zero following

koddsson committed:
trigger tests

koddsson committed:
also print numbers the same in parser

koddsson committed:
trigger tests

koddsson committed:
add handling string literals

koddsson committed:
updated error throwing

koddsson committed:
trigger tests

koddsson committed:
format groups correctly

koddsson committed:
check correct token in error

koddsson committed:
set correct exit code

koddsson committed:
fix grouping parse errors

koddsson committed:
fix bug in unary expressions

koddsson committed:
implement display for unary expressions

koddsson committed:
format binary operations and expressions correctly

koddsson commented:
All these commits are from pushing to my codecrafters.io origin to run unit tests against this code 😄

koddsson committed:
remove unneeded lifetimes

koddsson committed:
remove useless format! calls

koddsson committed:
fix more error messages

koddsson commented:
The codecrafters.io course is finished with these commits as all the tests pass which is cool!

But I feel like there's more in the book and actually stuff that the course didn't check in their tests like synchronising after errors and I believe that the "identifiers" codepath in my code is just unused.

So! Writing my own tests and then hardening sounds like a good idea before moving on to evaluating expressions which I believe is covered in the book.

koddsson committed:
remove `match_types` in favor of specialized functions

koddsson committed:
fix bug in match_equality_token

keithamus commented on src/expr.rs:
```diff
@@ -0,0 +1,97 @@
+
+#[derive(Debug, Clone)]
+pub enum Literal {
+    Str(String),
```
Again, here you'd ideally not copy `String`s over. Optimizing `Literal` is probably important because it's going to be used 1000s of times in a program, and doing heap allocations 1000s of times isn't great for performance.

Depending on how deep you want to go with this parser, here are some strategies to think about memory:

### General advice for `String`

So the "general rust advice" would be that this could be `&'a str` but then `Literal` will need a lifetime, which means `Expr` will and the whole thing bubbles up into a bit of a mess. `&str` is also 16 bytes, which means `Literal` will be at-least 16 bytes.

But potentially we could store less...

### Parser specific advice

Another way to think this is to think about the specific modelling of a parser itself. `Literal` does not sit in a vacuum. During the execution of your parser, the parser will be given the source to interpret, and the Parser will hold that. Think about what `&str` is and what data it might contain; if you were to write it as a struct:

```rust
struct Str {
  pointer_to_offset_of_real_string_in_memory: usize,
  length_of_this_slice_of_the_real_string: usize
}
```

There are two things that make this interesting, and why a specialisation might be worth pursuing in this one case. Firstly, `pointer_to_offset_of_real_string_in_memory` will point to a memory location where the underlying `String` is, at an offset of the desired string. Forgetting the offset part for a second, we don't necessarily _need_ to know where the `String` is in memory because the `Parser` has that, and `Literal` only really gets used in the context of the parser. So `Literal::Str` could be:

```rust
enum Literal {
  Str({ offset: usize, length: usize })
}
```

But we haven't gained anything here. This is still a 16byte enum. However let's think about `usize`. A `usize` is 8 bytes on 64bit systems, because it needs to be able to address up to ~18 Exabytes of memory (it's 4 bytes on 32-bit systems which means it can only address ~4gb of memory). I don't know about you, but I don't think I've ever encountered source code in the Gigabyte range, let alone Terra/Peta/Exabytes. So what's a realistic minimum? A `u8` is 1 byte but the max value is 255, and I've definitely seen a file bigger than 255 characters. A `u16`'s max value is `65535`, or 65MB... theoretically there are source files bigger than 65mb, especially if you're trying to parse big bundled source, but it would be very rare. And then we have the u32 which, again, is 4gb. 

So given these numbers, maybe the offset doesn't need to be `usize`, maybe it can be a `u32` and an appropriate limitation of your parser is that it won't parse files bigger than 4gb. Changing them both to `u32` slashes the struct from 16 to 8bytes:

```rust
enum Literal {
  Str({ offset: u32, length: u32 })
}
```

We _could_ apply the same technique to the `length` too; how likely is it that a program is going to _only_ have 1 string and that string is going to be 4gb? However, let's shrink `length` down to u16 and stick it in a file and run with Rust's `-Zprint-type-sizes` flag, to get size information:

```sh
$ cat file.rs
  #![allow(unused)]
  fn main() {
      enum Literal {
          Str(u32, u16),
      }
  }

$ rustc +nightly -Zprint-type-sizes file.rs

  ...
  print-type-size type: `main::Literal`: 8 bytes, alignment: 4 bytes
  print-type-size     variant `Str`: 6 bytes
  print-type-size         field `.0`: 4 bytes
  print-type-size         field `.1`: 2 bytes
  print-type-size     end padding: 2 bytes
  ...
```

Huh... so `field .0` (our u32) is 4 bytes. `field .1` is 2 bytes, and our `Str` variant is indeed 6 bytes. But the struct as a whole is 8 bytes, not 6... Rust added 2 bytes of padding!?

Rust will pad structs to specific memory boundaries, for efficiency in memory layout. It's just added 2 bytes of end padding to struct, making the whole thing 8 bytes. It tells us why: `alignment: 4 bytes`. It's trying to align the struct into 4 byte memory bounds. This means our optimization can't go lower than 4byte fields because Rust will just pad to multiples of 4.

So we've got two options:

1. We make `Str(u32, u16)` and live with the fact that Rust will pad it to 8 bytes anyway. This might save us later if we need to cram a couple bytes more data into this field, for example if we wanted to encode the quote style the string uses we could have `Str(u32, u16, u8)` and Rust would only pad the extra byte, meaning `Str(u32, u16, u8)` would still be 8 bytes.
2. We make `Str(u32, u32)` and move on with our lives.


Btw, for reference [here's a playground which demonstrates these sizes using `std::mem::size_of`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=20ad780d6ce84d4481481a0bbaaea57e).

koddsson replied:
> So the "general rust advice" would be that this could be `&'a str` but then `Literal` will need a lifetime, which means `Expr` will and the whole thing bubbles up into a bit of a mess.

This is what I was running into and _couldn't_ figure out what would be best. It's a good shout because I was gonna come back to this but forgot.

The comment is great (as always :pray:) I'll take a further look over the weekend!

koddsson replied:
Actually, one question: If I did all this, how do I take the offset and length and translate that to a string? Do I have to pass in a reference of the source code or something?

keithamus replied:
Yeah you'd have to have a method, perhaps on parser:

```rust
fn get_source_at_position(&self, start: u32, end: u32) -> &'a str {
  self.source[start..end]
}
```

koddsson replied:
What if the parser just has `Vec<Token>` and not actually the source code? 👀 

Should the parser have access to the actual source code?

keithamus replied:
Why would the parser end up with tokens but no source?

koddsson replied:
That's what the book walks you through. First you run the tokenizer on the source code and then you pass those tokens to the parser to get expressions.

https://craftinginterpreters.com/parsing-expressions.html#the-parser-class

keithamus replied:
Ah, okay. In that case the Tokenizer could house the source code instead.

koddsson replied:
Right, and I think it does right now? As a `&'a str`. Is it just a matter or "ordering" things? Instead of running the tokenizer and then the parser with it's output I should pass the whole tokenizer to the parser? Or have some sort of wrapping struct called compiler? Something that would allow the parser access to the source code (probably via the tokenizer).

keithamus replied:
For hdx the parser and tokenizer both get `&'a str` (remember that’s a cheap pointer so fine to pass around). But the tokenizer only lives as long as reading the source and Tokens themselves have no lifetimes bound, then the parser sticks around to build the AST which also has the same lifetime

koddsson replied:
Makes sense! I think for now I'm gonna add a TODO and start working on the interpreter and statements as I started working on that and got excited.

So for now I'm gonna merge this and create a new PR for the interpreter.

koddsson replied:
```suggestion
    // TODO: Convert this to a `&'a str`. See https://github.com/koddsson/codecrafters-interpreter-rust/pull/1#discussion_r1684154612
    Str(String),
```

koddsson committed:
remove a String from token_type

koddsson committed:
remove more String for &str

koddsson committed:
Call it a tokenzier and not a scanner
I realize that the "Crafting Interpreters" book calls it a scanner but
I feel like calling it tokenizer makes a lot more sense so I'm gonna
call it that.

koddsson committed:
Add TODO comment
